lib/linked_list.rb

```diff
@@ -7,39 +7,76 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
-  def add_first(data)
+  # Time complexity -  O(1)
+  # Space complexity -  O(1) only creating one node
+  def add_first(data) # creat a new node & assign it to @head
+    @head = Node.new(data, @head)
 
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
+    return nil if @head == nil
+    return @head.data # is always pointing to the first
 
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def length
-    return 0
+    return 0 if @head == nil
+    count = 1
+    current = head
+    while current.next != nil
+      current = current.next
+      count += 1
+    end
+    return count
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def add_last(data)
+    return add_first(data) if @head == nil
 
+    current = @head
+    while current.next != nil
+      current = current.next
+    end
+    current.next = Node.new(data, nil)
+    self
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def get_last
+    return nil if @head == nil
 
+    current = @head
+    while current.next != nil
+      current = current.next
+    end
+   return current.data
+  
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(n)
   def get_at_index(index)
```

Comment:
Question:  Why do you think there is O(n) space complexity?  You are not doing recursion or creating a new list.